File: LeetCode/Python/160.intersection-of-two-linked-lists.py
#
# @lc app=leetcode id=160 lang=python3
#
# [160] Intersection of Two Linked Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# TODO: Improve
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        ListA = [headA]
        ListB = [headB]

        tempHeadA = headA
        tempHeadB = headB

        while tempHeadA:
            tempHeadA = tempHeadA.next
            ListA.append(tempHeadA)

        while tempHeadB:
            tempHeadB = tempHeadB.next
            ListB.append(tempHeadB)
        
        x = None
        for (a, b) in list(zip(ListA[::-1], ListB[::-1])):
            if a == b:
                x = a
            else:
                return x

        return x

         
# @lc code=end

File: LeetCode/CPP/160.intersection-of-two-linked-lists.cpp
/*
 * @lc app=leetcode id=160 lang=cpp
 *
 * [160] Intersection of Two Linked Lists
 */
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    // HINT!- There is no performance improvement here, but the while loop can improved...
    // having true as the while loop condition is not a good practice
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB)
    {
        ListNode *A = headA, *B = headB;
        bool restartedA = false, restartedB = false;

        while (true)
        {
            if (A == nullptr)
            {
                A = headB;
                if (restartedA)
                    break;
                restartedA = true;
            }
            if (B == nullptr)
            {
                B = headA;
                if (restartedB)
                    break;
                restartedB = true;
            }
            if (A != nullptr && A == B)
                return A;
            A = A->next;
            B = B->next;
        }

        return nullptr;
    }
};
// @lc code=end
